File: ainpc-core-plugin/src/main/java/ro/ainpc/progression/ProgressionSelector.java
package ro.ainpc.progression;

import java.util.Locale;

public record ProgressionSelector(
    String raw,
    String normalized,
    String packId,
    String mechanicOrKind,
    String definitionId,
    boolean trackedAlias
) {
    public ProgressionSelector {
        raw = raw == null ? "" : raw.trim();
        normalized = normalized == null ? "" : normalized.trim();
        packId = packId == null ? "" : packId.trim();
        mechanicOrKind = mechanicOrKind == null ? "" : mechanicOrKind.trim();
        definitionId = definitionId == null ? "" : definitionId.trim();
    }

    public static ProgressionSelector parse(String selector) {
        String raw = selector == null ? "" : selector.trim();
        if (raw.isBlank()) {
            return new ProgressionSelector("", "", "", "", "", false);
        }

        String lower = raw.toLowerCase(Locale.ROOT);
        if (isTrackedAlias(lower)) {
            return new ProgressionSelector(raw, "tracked", "", "", "tracked", true);
        }

        String[] parts = raw.split(":", -1);
        if (parts.length == 2) {
            String mechanicOrKind = parts[0].trim();
            String definitionId = parts[1].trim();
            return new ProgressionSelector(raw, mechanicOrKind + ":" + definitionId, "", mechanicOrKind, definitionId, false);
        }
        if (parts.length == 3) {
            String packId = parts[0].trim();
            String mechanicOrKind = parts[1].trim();
            String definitionId = parts[2].trim();
            return new ProgressionSelector(raw, packId + ":" + mechanicOrKind + ":" + definitionId, packId, mechanicOrKind, definitionId, false);
        }

        return new ProgressionSelector(raw, raw, "", "", raw, false);
    }

    public static ProgressionSelector forContractAlias(String selector) {
        return forKindAlias(selector, "contract");
    }

    public static ProgressionSelector forKindAlias(String selector, String kind) {
        ProgressionSelector parsed = parse(selector);
        if (parsed.isEmpty()
            || parsed.hasNamespace()
            || parsed.isTrackedAlias()
            || "nearest".equalsIgnoreCase(parsed.normalized())) {
            return parsed;
        }

        String safeKind = kind == null ? "" : kind.trim();
        if (safeKind.isBlank()) {
            return parsed;
        }
        return parse(safeKind + ":" + parsed.normalized());
    }

    public static boolean isTrackedAlias(String selector) {
        String normalized = selector == null ? "" : selector.trim().toLowerCase(Locale.ROOT);
        return "tracked".equals(normalized)
            || "current".equals(normalized)
            || "curent".equals(normalized)
            || "urmarit".equals(normalized);
    }

    public boolean isEmpty() {
        return normalized.isBlank();
    }

    public boolean hasNamespace() {
        return !mechanicOrKind.isBlank() || !packId.isBlank();
    }

    public boolean isTrackedAlias() {
        return trackedAlias;
    }

    public String commandSelector() {
        return normalized;
    }
}

File: ainpc-core-plugin/bin/main/ro/ainpc/progression/ProgressionSelector.kt
package ro.ainpc.progression

import java.util.Locale
import java.util.regex.Pattern

class ProgressionSelector(
    raw: String?,
    normalized: String?,
    packId: String?,
    mechanicOrKind: String?,
    definitionId: String?,
    trackedAlias: Boolean
) {
    private val rawValue = valueOrEmpty(raw)
    private val normalizedValue = valueOrEmpty(normalized)
    private val packIdValue = valueOrEmpty(packId)
    private val mechanicOrKindValue = valueOrEmpty(mechanicOrKind)
    private val definitionIdValue = valueOrEmpty(definitionId)
    private val trackedAliasValue = trackedAlias

    fun raw(): String = rawValue

    fun normalized(): String = normalizedValue

    fun packId(): String = packIdValue

    fun mechanicOrKind(): String = mechanicOrKindValue

    fun definitionId(): String = definitionIdValue

    fun trackedAlias(): Boolean = trackedAliasValue

    fun isEmpty(): Boolean = normalizedValue.isBlank()

    fun hasNamespace(): Boolean = mechanicOrKindValue.isNotBlank() || packIdValue.isNotBlank()

    fun isTrackedAlias(): Boolean = trackedAliasValue

    fun commandSelector(): String = normalizedValue

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is ProgressionSelector) {
            return false
        }

        return rawValue == other.rawValue &&
            normalizedValue == other.normalizedValue &&
            packIdValue == other.packIdValue &&
            mechanicOrKindValue == other.mechanicOrKindValue &&
            definitionIdValue == other.definitionIdValue &&
            trackedAliasValue == other.trackedAliasValue
    }

    override fun hashCode(): Int {
        var result = rawValue.hashCode()
        result = 31 * result + normalizedValue.hashCode()
        result = 31 * result + packIdValue.hashCode()
        result = 31 * result + mechanicOrKindValue.hashCode()
        result = 31 * result + definitionIdValue.hashCode()
        result = 31 * result + trackedAliasValue.hashCode()
        return result
    }

    override fun toString(): String =
        "ProgressionSelector[raw=$rawValue, normalized=$normalizedValue, " +
            "packId=$packIdValue, mechanicOrKind=$mechanicOrKindValue, " +
            "definitionId=$definitionIdValue, trackedAlias=$trackedAliasValue]"

    companion object {
        private val SELECTOR_SEPARATOR: Pattern = Pattern.compile(":")

        @JvmStatic
        fun parse(selector: String?): ProgressionSelector {
            val raw = valueOrEmpty(selector)
            if (raw.isBlank()) {
                return ProgressionSelector("", "", "", "", "", false)
            }

            val lower = raw.lowercase(Locale.ROOT)
            if (isTrackedAlias(lower)) {
                return ProgressionSelector(raw, "tracked", "", "", "tracked", true)
            }

            val parts = SELECTOR_SEPARATOR.split(raw, -1)
            if (parts.size == 2) {
                val mechanicOrKind = parts[0].trim()
                val definitionId = parts[1].trim()
                return ProgressionSelector(
                    raw,
                    "$mechanicOrKind:$definitionId",
                    "",
                    mechanicOrKind,
                    definitionId,
                    false
                )
            }
            if (parts.size == 3) {
                val packId = parts[0].trim()
                val mechanicOrKind = parts[1].trim()
                val definitionId = parts[2].trim()
                return ProgressionSelector(
                    raw,
                    "$packId:$mechanicOrKind:$definitionId",
                    packId,
                    mechanicOrKind,
                    definitionId,
                    false
                )
            }

            return ProgressionSelector(raw, raw, "", "", raw, false)
        }

        @JvmStatic
        fun forContractAlias(selector: String?): ProgressionSelector = forKindAlias(selector, "contract")

        @JvmStatic
        fun forKindAlias(selector: String?, kind: String?): ProgressionSelector {
            val parsed = parse(selector)
            if (parsed.isEmpty() ||
                parsed.hasNamespace() ||
                parsed.isTrackedAlias() ||
                "nearest".equals(parsed.normalized(), ignoreCase = true)
            ) {
                return parsed
            }

            val safeKind = valueOrEmpty(kind)
            if (safeKind.isBlank()) {
                return parsed
            }
            return parse("$safeKind:${parsed.normalized()}")
        }

        @JvmStatic
        fun isTrackedAlias(selector: String?): Boolean {
            val normalized = valueOrEmpty(selector).lowercase(Locale.ROOT)
            return normalized == "tracked" ||
                normalized == "current" ||
                normalized == "curent" ||
                normalized == "urmarit"
        }

        private fun valueOrEmpty(value: String?): String = value?.trim().orEmpty()
    }
}
